File: frontend/js/views/offer_letter.js
// Employment offer letter generator — per-paycheck breakdown + offer clauses,
// via /calc/offer-letter. Previews live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderOfferLetter(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.offer.h1.title">// EMPLOYMENT OFFER LETTER</span></h1>
        <p class="muted small" data-i18n="view.offer.hint.intro">
            Extends a job offer with title, compensation, and start date. It breaks the annual salary into
            the per-paycheck amount for the chosen pay frequency and assembles the offer clauses
            (position, compensation, equity, benefits, at-will, contingencies). Drafting aid, not legal
            advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.offer.h2.inputs">Offer details</h2>
            <form id="offer-form" class="inline-form">
                <label><span data-i18n="view.offer.label.company">Company</span>
                    <input type="text" name="company_name" value=""></label>
                <label><span data-i18n="view.offer.label.candidate">Candidate</span>
                    <input type="text" name="candidate_name" value=""></label>
                <label><span data-i18n="view.offer.label.title">Job title</span>
                    <input type="text" name="job_title" value="Software Engineer"></label>
                <label><span data-i18n="view.offer.label.salary">Annual salary ($)</span>
                    <input type="number" step="100" min="0" name="annual_salary_usd" value="120000" required></label>
                <label><span data-i18n="view.offer.label.freq">Pay frequency</span>
                    <select name="pay_frequency">
                        <option value="weekly" data-i18n="view.offer.freq.weekly">Weekly</option>
                        <option value="biweekly" selected data-i18n="view.offer.freq.biweekly">Biweekly</option>
                        <option value="semimonthly" data-i18n="view.offer.freq.semimonthly">Semimonthly</option>
                        <option value="monthly" data-i18n="view.offer.freq.monthly">Monthly</option>
                    </select></label>
                <label><span data-i18n="view.offer.label.start">Start date</span>
                    <input type="date" name="start_date" value="2026-08-01" required></label>
                <label><span data-i18n="view.offer.label.bonus">Signing bonus ($)</span>
                    <input type="number" step="100" min="0" name="signing_bonus_usd" value="10000"></label>
                <label><span data-i18n="view.offer.label.equity">Equity (optional)</span>
                    <input type="text" name="equity_description" value="5,000 ISOs vesting over 4 years"></label>
                <label><span data-i18n="view.offer.label.exempt">FLSA-exempt (salaried)</span>
                    <input type="checkbox" name="exempt" checked></label>
                <label><span data-i18n="view.offer.label.state">State</span>
                    <input type="text" name="state" value="California" required></label>
            </form>
        </div>
        <div id="offer-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#offer-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            company_name: (fd.get('company_name') || '').trim(),
            candidate_name: (fd.get('candidate_name') || '').trim(),
            job_title: (fd.get('job_title') || '').trim(),
            annual_salary_usd: Number(fd.get('annual_salary_usd')) || 0,
            pay_frequency: fd.get('pay_frequency'),
            start_date: fd.get('start_date'),
            signing_bonus_usd: Number(fd.get('signing_bonus_usd')) || 0,
            equity_description: (fd.get('equity_description') || '').trim(),
            exempt: fd.get('exempt') != null,
            state: (fd.get('state') || '').trim(),
        };
        try {
            const doc = await api.calcOfferLetter(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.offer.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#offer-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.offer.card.paycheck">Per paycheck</div>
                    <div class="value">${money(doc.per_paycheck_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.offer.card.salary">Annual salary</div>
                    <div class="value">${money(doc.annual_salary_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.offer.card.periods">Pay periods/yr</div>
                    <div class="value">${doc.periods_per_year}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="offer-copy" type="button" data-i18n="view.offer.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="offer-download" type="button" data-i18n="view.offer.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#offer-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.offer.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.offer.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#offer-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'offer-letter.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/offer_letter.rs
//! Employment offer letter — extends a job offer with title, compensation, and
//! start date. It breaks the annual salary into the per-paycheck amount for the
//! chosen pay frequency and assembles the offer clauses (position, compensation,
//! equity, benefits, at-will, contingencies). Drafting aid, not legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PayFrequency {
    Weekly,
    Biweekly,
    Semimonthly,
    Monthly,
}

impl PayFrequency {
    fn periods_per_year(self) -> i32 {
        match self {
            PayFrequency::Weekly => 52,
            PayFrequency::Biweekly => 26,
            PayFrequency::Semimonthly => 24,
            PayFrequency::Monthly => 12,
        }
    }
    fn label(self) -> &'static str {
        match self {
            PayFrequency::Weekly => "weekly",
            PayFrequency::Biweekly => "biweekly",
            PayFrequency::Semimonthly => "semimonthly",
            PayFrequency::Monthly => "monthly",
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct OfferInput {
    pub company_name: String,
    pub candidate_name: String,
    pub job_title: String,
    pub annual_salary_usd: f64,
    pub pay_frequency: PayFrequency,
    pub start_date: String,
    #[serde(default)]
    pub signing_bonus_usd: f64,
    #[serde(default)]
    pub equity_description: String,
    /// FLSA-exempt (salaried) vs non-exempt (overtime-eligible).
    #[serde(default)]
    pub exempt: bool,
    pub state: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct OfferLetter {
    pub title: String,
    pub annual_salary_usd: f64,
    pub periods_per_year: i32,
    pub per_paycheck_usd: f64,
    pub signing_bonus_usd: f64,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &OfferInput) -> OfferLetter {
    let ppy = i.pay_frequency.periods_per_year();
    let per_paycheck = cents(i.annual_salary_usd / ppy as f64);

    let bonus_part = if i.signing_bonus_usd > 0.0 {
        format!(" You will also receive a one-time signing bonus of {}.", money(i.signing_bonus_usd))
    } else {
        String::new()
    };

    let exempt_part = if i.exempt {
        " This is a salaried, exempt position and is not eligible for overtime."
    } else {
        " This is a non-exempt position and is eligible for overtime as required by law."
    };

    let clauses = vec![
        DocClause {
            heading: "1. Position".into(),
            body: format!(
                "{} is pleased to offer {} the position of {}, beginning {}.{}",
                i.company_name, i.candidate_name, i.job_title, i.start_date, exempt_part
            ),
        },
        DocClause {
            heading: "2. Compensation".into(),
            body: format!(
                "Your annual base salary is {}, paid {} ({} per paycheck across {} pay periods per year).{}",
                money(i.annual_salary_usd),
                i.pay_frequency.label(),
                money(per_paycheck),
                ppy,
                bonus_part
            ),
        },
        DocClause {
            heading: "3. Equity".into(),
            body: if i.equity_description.trim().is_empty() {
                "No equity is included in this offer.".to_string()
            } else {
                format!("Subject to board approval and the company's plan, you will be granted: {}.", i.equity_description.trim())
            },
        },
        DocClause {
            heading: "4. Benefits".into(),
            body: "You will be eligible to participate in the company's standard benefit programs (health, retirement, paid time off) in accordance with their terms.".into(),
        },
        DocClause {
            heading: "5. At-Will Employment".into(),
            body: format!(
                "Your employment with {} is at-will under the laws of the State of {}, meaning either you or the company may terminate the employment relationship at any time, with or without cause or notice.",
                i.company_name, i.state
            ),
        },
        DocClause {
            heading: "6. Contingencies".into(),
            body: "This offer is contingent on your eligibility to work in the United States (Form I-9 verification) and the satisfactory completion of any background check the company requires.".into(),
        },
        DocClause {
            heading: "Acceptance".into(),
            body: format!(
                "Candidate signature: ____________________  Date: __________\n{}\n\nCompany representative: ____________________  Date: __________\n{}",
                i.candidate_name, i.company_name
            ),
        },
    ];

    OfferLetter {
        title: "Employment Offer Letter".into(),
        annual_salary_usd: i.annual_salary_usd,
        periods_per_year: ppy,
        per_paycheck_usd: per_paycheck,
        signing_bonus_usd: i.signing_bonus_usd,
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> OfferInput {
        OfferInput {
            company_name: "Acme Inc".into(),
            candidate_name: "Jordan Hire".into(),
            job_title: "Software Engineer".into(),
            annual_salary_usd: 120_000.0,
            pay_frequency: PayFrequency::Biweekly,
            start_date: "2026-08-01".into(),
            signing_bonus_usd: 10_000.0,
            equity_description: "5,000 ISOs vesting over 4 years".into(),
            exempt: true,
            state: "California".into(),
        }
    }

    #[test]
    fn biweekly_per_paycheck() {
        let d = generate(&base());
        assert_eq!(d.periods_per_year, 26);
        assert!(close(d.per_paycheck_usd, 4_615.38));
    }

    #[test]
    fn weekly_and_monthly() {
        assert!(close(generate(&OfferInput { pay_frequency: PayFrequency::Weekly, ..base() }).per_paycheck_usd, 2_307.69));
        assert!(close(generate(&OfferInput { pay_frequency: PayFrequency::Monthly, ..base() }).per_paycheck_usd, 10_000.0));
        assert!(close(generate(&OfferInput { pay_frequency: PayFrequency::Semimonthly, ..base() }).per_paycheck_usd, 5_000.0));
    }

    #[test]
    fn signing_bonus_in_clause() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading == "2. Compensation").unwrap();
        assert!(c.body.contains("signing bonus of $10000.00"));
    }

    #[test]
    fn no_bonus_omits_it() {
        let c = generate(&OfferInput { signing_bonus_usd: 0.0, ..base() })
            .clauses.into_iter().find(|c| c.heading == "2. Compensation").unwrap();
        assert!(!c.body.contains("signing bonus"));
    }

    #[test]
    fn exempt_vs_nonexempt() {
        assert!(generate(&base()).clauses.iter().find(|c| c.heading == "1. Position").unwrap().body.contains("not eligible for overtime"));
        let ne = generate(&OfferInput { exempt: false, ..base() });
        assert!(ne.clauses.iter().find(|c| c.heading == "1. Position").unwrap().body.contains("eligible for overtime"));
    }

    #[test]
    fn equity_clause_optional() {
        assert!(generate(&base()).clauses.iter().find(|c| c.heading == "3. Equity").unwrap().body.contains("5,000 ISOs"));
        let none = generate(&OfferInput { equity_description: String::new(), ..base() });
        assert!(none.clauses.iter().find(|c| c.heading == "3. Equity").unwrap().body.contains("No equity"));
    }

    #[test]
    fn at_will_names_state() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading.contains("At-Will")).unwrap();
        assert!(c.body.contains("California"));
        assert!(c.body.contains("at-will"));
    }
}
